rfc/ipni-w3c.md

```diff
@@ -0,0 +1,346 @@
+    "location": "https://r2.cf/bag...carx"
+  }
+}
+```
+
+**Inclusion claim** example
+
+```json
+{
+  "content": CID, // CAR CID
+  "includes": CID // CARx-index Index CID
+}
+```
+
+When `ipni/offer` is invoked, encoded inclusion claim and encoded location claims for the CAR and the CARx-index are all sent with the invocation.
+
+_Alternatively, the w3s service could create the location claims. The user would still need to provide the locations, so might as well do this as a location claim. If w3s moves the data, then it can replace the location claims with new ones_
+
+The service must fetch he CARx-index and parse it to find the set of multihashes included in the CAR. see: [Verifying the CARv2 Index](#verifying-the-carv2-index)
+
+_Alternatively the service could generate the CARx-index and store it, but it would be better to have the clients do that work of generating and storing the index._
+
+Questions:
+1. Is another claim needed to identify the CID of each location claim?
+2. Will there be a content claims service where clients can lookup content claims using the CID of the CAR file or some key associated with the CAR file? This RFC assumes that the content claims bucket is accessible via URL which can either be a direct file reference or a be a service endpoint.
+
+
+#### Verifying the CARx-index Index
+
+The service must fetch the CARx-index and may verify 1 or more multihashes from the index exist at the specified offsets in the associated CAR.
+
+The verifier should pick a set of multihashes at random and fetch the bytes from the CAR identified by the index entry and verify it's multihash. The invocation must return an error if any entry is found to be invalid.
+
+Random validation of a number of blocks allows us to detect invalid indexes and lets us tune how much work we are willing to do per car index.
+
+Full validation of every block is not recommended as it opens us up to performing unbounded work. _We have seen CAR files with millions of tiny blocks._
+#### Advertisement construction
+
+The set of multihashes must be encoded as 1 [IPNI Advertisement], linking to 1 or more [Entries] per the IPLD Schema:
+
+```ipldsch
+type Advertisement struct {
+    PreviousID optional Link
+    Provider String
+    Addresses [String]
+    Signature Bytes
+    Entries Link
+    ContextID Bytes
+    Metadata Bytes
+    IsRm Bool
+    ExtendedProvider optional ExtendedProvider
+}
+
+type EntryChunk struct {
+    Entries [Bytes]
+    Next optional Link
+}
+```
+- `Addresses` the most recent value of this is returned with lookups for any CID in any advertisement. It will contain the multiaddr form of URL claims service? TBD.
+- `ContextID` must be the byte encoded form of the CAR CID.
+- `Entries` must be the CID of an `EntryChunk` for a subset (or all) of the multihashes in the CAR.
+- `Metadata` encodes the location of the content claims bucket and the inclusion claim that ties the CAR CID to the CARx-index CID. 
```

Comment:
I think it would be better to not put any content claim here, but rather just the minimal needed to get the content claims bucket. 

The advertisement is signed so it cannot be altered without invalidating the signature/CID, so if the ad contains the CID of the claims bucket and the claims bucket has an inclusion claim that contains the advertisement CID, then that should provide a chain of provable authenticity.

Reply:
I agree. As you know I think it might make sense to put a location claim for the car-index in here for parallelism of round trips.

Reply:
Truth be told this will all work better though if we can get away from car files one day :P

Reply:
But even if we do, I assume we will be stuck supporting car files in perpetuity.